### PR TITLE
Make ICG tests pass when GPDB is compiled with disable-orca

### PR DESCRIPTION
**Pull request by vraghavan78**
In the updated tests, we used functions like disable_xform and
enable_xform to hint the optimizer to disallow/allow a particular
physical node. However, these functions are only available when GPDB
is built with GPORCA. Planner on the other hand accomplished this
via a GUC.

To avoid usage of these functions in tests, I have introduced couple
of GUCS that mimic the same planner behavior but now for GPORCA.
In this effort I needed to add an API inside GPORCA.

https://github.com/greenplum-db/gporca/pull/177

By using these gucs accompanied by the changes in the sql/out files the
following tests now pass when we build GPDB with --disable-orca

* bfv_catalog
* bfv_partition
* bitmapscan_ao
* bitmapscan
* indexjoin
* qp_left_anti_semi_join
* bfv_statistic

In the process I cleaned up:
* bfv_index
* co_nestloop_idxscan
* join_gp

I am working on the remaining failing tests:
* gp_optimizer
* gp_metadata

### CODE REVIEW

**Comment by vraghavan78** [comment]
Folks, I would like to fix the remaining failing tests in another PR.

File: src/test/regress/expected/bfv_statistic.out [hunk 1]
```diff
@@ -26,12 +16,16 @@ select stanullfrac, stadistinct, stanumbers1 from pg_statistic where starelid='f
         0.25 |          -1 | 
 (1 row)
 
--- exercise the translator
-select check_row_count('explain select * from foo where a is not null and b >= 1;', 3);
- check_row_count 
------------------
- true
-(1 row)
+-- exercise GPORCA translator
```
Comment (hlinnaka) [inline src/test/regress/expected/bfv_statistic.out]: I don't understand the purpose of this test query. It's a fairly simple query, and the table doesn't have any indexes, so you're going to get the same plan regardless of the statistics.

File: src/test/regress/expected/bfv_statistic.out [hunk 2]
```diff
@@ -3,16 +3,6 @@ set search_path=bfv_statistic;
 -- start_ignore
 create language plpythonu;
```
Comment (hlinnaka) [inline src/test/regress/expected/bfv_statistic.out]: No need for plpython anymore AFAICS

File: src/test/regress/expected/bfv_statistic.out [hunk 3]
```diff
@@ -48,21 +42,35 @@ select stanumbers1, stavalues1 from pg_statistic where starelid='foo2'::regclass
...
 update pg_statistic set stavalues1='{6,3,1,5,4,2}'::int[] where starelid='foo2'::regclass;
 -- excercise the translator
-select check_row_count('explain select a from foo2 where a > 1 order by a;', 8);
- check_row_count 
------------------
- true
-(1 row)
+explain select a from foo2 where a > 1 order by a;
+                                 QUERY PLAN                                  
+-----------------------------------------------------------------------------
```
Comment (hlinnaka) [inline src/test/regress/expected/bfv_statistic.out]: Same here and in the previous query, the statistics won't make any difference.

File: src/test/regress/expected/bfv_statistic.out [hunk 4]
```diff
@@ -154,61 +163,24 @@ select v from (select max(c1) as v, 1 as r from t1 union select 1 as v, 2 as r )
...
- CXformPushGbBelowUnionAll is enabled
-(1 row)
-
-select disable_xform('CXformPushGbBelowUnion');
-           disable_xform            
-------------------------------------
- CXformPushGbBelowUnion is disabled
-(1 row)
-
 select v from (select max(c1) as v, 1 as r from t1 union select 1 as v, 2 as r ) as foo group by v;
```
Comment (hlinnaka) [inline src/test/regress/expected/bfv_statistic.out]: This query is identical to the previous one.

File: src/backend/utils/misc/guc_gp.c [hunk 5]
```diff
@@ -2785,6 +2788,36 @@ struct config_bool ConfigureNamesBool_gp[] =
...
+			gettext_noop("Enables the optimizer's use of plans with dynamic table scan."),
+			NULL,
+			GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE
+		},
+		&optimizer_enable_dynamic_table_scan,
+		true, NULL, NULL
+	},
+
+	{
+		{"optimizer_enable_index_scan", PGC_USERSET, DEVELOPER_OPTIONS,
```
Comment (hlinnaka) [inline src/backend/utils/misc/guc_gp.c]: Can we call this `optimizer_enable_indexscan`, to mirror the planner's `enable_indexscan`?

File: src/test/regress/expected/co_nestloop_idxscan_1.out [hunk 6]
```diff
@@ -28,12 +28,7 @@ select f.id from co_nestloop_idxscan.foo f, co_nestloop_idxscan.bar b where f.id
 (1 row)
 
 -- test with nested loop join
-select disable_xform('CXformInnerJoin2HashJoin');
```
Comment (hlinnaka) [inline src/test/regress/expected/co_nestloop_idxscan_1.out]: I think you can remove this alternative expected output file altogether now. It was added in commit 0ac120aadd to make the test pass with --disable-orca, but now that it doesn't call disable_xform anymore, it shouldn't be required.

File: src/test/regress/expected/join_gp_1.out [hunk 7]
```diff
@@ -260,37 +260,7 @@ left outer join (tjoin2 left outer join tjoin3 on tjoin2.id=tjoin3.id) on tjoin1
 (12 rows)
 
 set enable_hashjoin to off;
--- Disable hashjoin forms fo ORCA
```
Comment (hlinnaka) [inline src/test/regress/expected/join_gp_1.out]: I believe this alternative expected output file can also be removed now.

**Comment by hlinnaka** [comment]
Thanks for working on this!

File: src/backend/utils/misc/guc_gp.c (hunk 5 above)
Comment (vraghavan78) [inline src/backend/utils/misc/guc_gp.c]: Sure

File: src/test/regress/expected/bfv_statistic.out (hunk 1 above)
Comment (vraghavan78) [inline src/test/regress/expected/bfv_statistic.out]: This test is badly named. It has to do with cardinality estimation inside orca. I can convert it into orca unit test

File: src/test/regress/expected/bfv_statistic.out (hunk 3 above)
Comment (vraghavan78) [inline src/test/regress/expected/bfv_statistic.out]: It is the cardinality estimation done by orca. I will convert it into orca unit test

**Comment by andreasscherbaum** [comment]
Also reported here: https://github.com/greenplum-db/gpdb/issues/2467

**Comment by andreasscherbaum** [comment]
I have the following two tests failing:

```
    gp_metadata          ... FAILED (0,01 sec)  (diff:0,10 sec)
     gp_optimizer         ... FAILED (20,08 sec)  (diff:0,96 sec)
```

**Comment by Eulerizeit** [comment]
Yep, those are known to still be failing Venky mentioned it the other day
when he was submitting the PR.

-- Rob

**Comment by andreasscherbaum** [comment]
Jepp, he asked for review.

File: src/test/regress/sql/bfv_statistic.sql [hunk 8]
```diff
@@ -123,19 +106,11 @@ select v from (select max(c1) as v, 1 as r from t1 union all select 1 as v, 2 as
...
-select disable_xform('CXformPushGbBelowUnionAll');
 select v from (select max(c1) as v, 1 as r from t1 union select 1 as v, 2 as r ) as foo group by v;
-select enable_xform('CXformPushGbBelowUnionAll');
 
-select disable_xform('CXformPushGbBelowUnion');
 select v from (select max(c1) as v, 1 as r from t1 union select 1 as v, 2 as r ) as foo group by v;
-select enable_xform('CXformPushGbBelowUnion');
 
-select disable_xform('CXformSimplifyGbAgg');
 select v from (select max(c1) as v, 1 as r from t1 union select 1 as v, 2 as r ) as foo group by v;
```
Comment (asimrp) [inline src/test/regress/sql/bfv_statistic.sql]: Heikki already pointed this out.  The query `select v from (select max(c1) as v, 1 as r from t1 union select 1 as v, 2 as r ) as foo group by v;` repeats 4 times in this sql file: lines 113, 111, 109 and 103.  As far as I can tell, running the query only once should suffice.  Not a fault of this patch but will be good to fix while at it.

Comment (vraghavan78) [inline src/test/regress/sql/bfv_statistic.sql]: will do.

**Comment by vraghavan78** [comment]
Thanks folks. Merged.

**Pull request closed without merging**